http-client/src/main/java/io/micronaut/http/client/netty/PoolResizer.java

```diff
@@ -250,11 +310,72 @@ final void markConnectionAvailable() {
     }
 
     final void forEachConnection(Consumer<ResizerConnection> c) {
-        for (ResizerConnection http1Connection : http1Connections) {
-            c.accept(http1Connection);
+        http1Connections.forEach(c);
+        http2Connections.forEach(c);
+    }
+
+    /**
+     * This is a concurrent list implementation that is similar to
+     * {@link java.util.concurrent.CopyOnWriteArrayList}, but with some extra optimization for
+     * {@link #doSomeWork()}.
+     */
+    private static final class ConnectionList {
+        private static final ResizerConnection[] EMPTY = new ResizerConnection[0];
+
+        private final Lock lock = new ReentrantLock();
+
+        /**
+         * Copy of {@link #safeItems} <i>only</i> for use in {@link #doSomeWork()}, without lock.
+         * {@link #doSomeWork()} may shuffle and reorder this array in-place as needed.
+         */
+        private volatile ResizerConnection[] unsafeItems = EMPTY;
```

Comment:
Just use `CopyOnWriteArrayList`

Reply:
i want to avoid lock contention in doSomeWork, and it also gives me the option to sort the connections in-place there